File: lua/copilot/define_tool.lua
--- Helper for defining tools with automatic result normalization.
--
-- Usage:
--   local define_tool = require("copilot.define_tool")
--
--   local weather_tool = define_tool("get_weather", {
--       description = "Get weather for a city",
--       parameters = {
--           type = "object",
--           properties = {
--               city = { type = "string", description = "City name" },
--               unit = { type = "string", description = "celsius or fahrenheit" },
--           },
--           required = { "city" },
--       },
--       handler = function(args, invocation)
--           return string.format("Weather in %s: 22 degrees %s",
--               args.city or "unknown", args.unit or "celsius")
--       end,
--   })

local cjson = require("cjson")
local types = require("copilot.types")

--- Normalize a tool handler return value into a ToolResult table.
-- Strings pass through directly, ToolResult tables pass through,
-- other values are JSON-serialized.
-- @param result any  The raw return value from the handler
-- @return table ToolResult
local function normalize_result(result)
    if result == nil then
        return types.ToolResult({
            textResultForLlm = "",
            resultType       = "success",
        })
    end

    -- If it is already a ToolResult-shaped table (has resultType field)
    if type(result) == "table" and result.resultType then
        return result
    end

    -- Strings pass through directly
    if type(result) == "string" then
        return types.ToolResult({
            textResultForLlm = result,
            resultType       = "success",
        })
    end

    -- Everything else gets JSON-serialized
    local ok, json_str = pcall(cjson.encode, result)
    if not ok then
        return types.ToolResult({
            textResultForLlm = tostring(result),
            resultType       = "success",
        })
    end

    return types.ToolResult({
        textResultForLlm = json_str,
        resultType       = "success",
    })
end

--- Define a tool with automatic result normalization.
--
-- The handler receives the parsed arguments table and a ToolInvocation table.
-- It can return:
--   - A string (becomes textResultForLlm)
--   - A ToolResult table (passed through as-is)
--   - Any other value (JSON-serialized to textResultForLlm)
--   - nil (empty success result)
--
-- If the handler raises an error, a failure ToolResult is returned automatically.
--
-- @param name string             Tool name
-- @param opts table with fields:
--   - description: string        Tool description
--   - parameters: table|nil      JSON Schema for arguments
--   - handler: function(args, invocation) -> any
-- @return table Tool              A Tool table ready for use in SessionConfig
local function define_tool(name, opts)
    opts = opts or {}

    local raw_handler = opts.handler
    if not raw_handler then
        error("define_tool: handler is required")
    end

    -- Wrap the handler to normalize results and catch errors
    local wrapped_handler = function(invocation)
        -- Parse arguments if they are a string (defensive)
        local args = invocation.arguments
        if type(args) == "string" then
            local ok, parsed = pcall(cjson.decode, args)
            if ok then
                args = parsed
            end
        end
        args = args or {}

        -- Call the user handler
        local ok, result_or_err = pcall(raw_handler, args, invocation)
        if not ok then
            return types.ToolResult({
                textResultForLlm = "Invoking this tool produced an error. Detailed information is not available.",
                resultType       = "failure",
                error            = tostring(result_or_err),
                toolTelemetry    = {},
            })
        end

        return normalize_result(result_or_err)
    end

    return types.Tool({
        name        = name,
        description = opts.description or "",
        parameters  = opts.parameters,
        handler     = wrapped_handler,
    })
end

return define_tool

File: clojure/src/copilot/define_tool.clj
;;;; ---------------------------------------------------------------------------
;;;;  Copyright (c) Microsoft Corporation. All rights reserved.
;;;; ---------------------------------------------------------------------------
;;;;
;;;; Helper macro and function for defining tools idiomatically in Clojure.

(ns copilot.define-tool
  "Idiomatic helpers for defining Copilot tools.

  Provides `define-tool` (function) and `deftool` (macro) for concise
  tool definitions with automatic result normalisation."
  (:require [cheshire.core :as json]
            [copilot.types :as types]))

;; ============================================================================
;; Result normalisation
;; ============================================================================

(defn- normalize-result
  "Convert any return value to a ToolResult map.

  - nil            -> empty success
  - String         -> success with text
  - ToolResultObject (map with :textResultForLlm + :resultType) -> pass through
  - Anything else  -> JSON-serialised success"
  [result]
  (cond
    (nil? result)
    {:textResultForLlm ""
     :resultType       "success"
     :toolTelemetry    {}}

    ;; Already a ToolResultObject
    (and (map? result)
         (contains? result :textResultForLlm)
         (contains? result :resultType))
    result

    ;; Plain string
    (string? result)
    {:textResultForLlm result
     :resultType       "success"
     :toolTelemetry    {}}

    ;; Everything else -> JSON
    :else
    {:textResultForLlm (json/generate-string result)
     :resultType       "success"
     :toolTelemetry    {}}))

;; ============================================================================
;; define-tool function
;; ============================================================================

(defn define-tool
  "Define a tool with automatic result normalisation.

  The handler function receives `(args invocation)` where:
    - `args`       is the parsed arguments map from the LLM
    - `invocation` is a map with :session-id :tool-call-id :tool-name :arguments

  The handler may return:
    - A string (wrapped as success)
    - A ToolResultObject map (passed through)
    - Any other value (JSON-serialised as success)
    - nil (empty success)

  Exceptions in the handler are caught and returned as failure results with
  the error detail hidden from the LLM.

  Parameters:
    `name`        - tool name string
    `description` - description shown to the LLM
    `parameters`  - JSON Schema map for the tool's parameters (or nil)
    `handler-fn`  - (fn [args invocation]) -> any

  Returns a tool map suitable for `copilot.client/create-session!` :tools."
  ([name description handler-fn]
   (define-tool name description nil handler-fn))
  ([name description parameters handler-fn]
   {:name        name
    :description (or description "")
    :parameters  parameters
    :handler     (fn [args invocation]
                   (try
                     (normalize-result (handler-fn args invocation))
                     (catch Exception e
                       {:textResultForLlm "Invoking this tool produced an error. Detailed information is not available."
                        :resultType       "failure"
                        :error            (.getMessage e)
                        :toolTelemetry    {}})))}))

;; ============================================================================
;; deftool macro
;; ============================================================================

(defmacro deftool
  "Define a tool and bind it to a var.

  Usage:

    (deftool lookup-fact
      \"Returns a fun fact about a given topic.\"
      {:type       \"object\"
       :properties {:topic {:type \"string\"
                            :description \"Topic to look up\"}}
       :required   [\"topic\"]}
      [args invocation]
      (get facts (get args :topic) \"No fact found.\"))

  Expands to a `def` binding to the result of `define-tool`.

  The body has access to the bindings specified in the parameter vector.
  The first binding receives the arguments map, the second (optional)
  receives the invocation context map."
  {:arglists '([name docstring schema [args-binding invocation-binding] & body]
               [name docstring [args-binding invocation-binding] & body])}
  [var-name description & more]
  (let [[schema bindings & body]
        (if (map? (first more))
          more
          (cons nil more))
        tool-name (clojure.core/name var-name)
        args-sym  (first bindings)
        inv-sym   (or (second bindings) (gensym "inv"))]
    `(def ~var-name
       (define-tool ~tool-name
                    ~description
                    ~schema
                    (fn [~args-sym ~inv-sym]
                      ~@body)))))
